### PR TITLE
Add Custom Downsampler

### LINKED ISSUE #973: Balanced sampling for classification problems
#971 tracks adding a data check to detect and warn about severe class-imbalance. #457 tracks adding an automl objective which can perform well on imbalanced datasets

Detecting and warning about class imbalance could be helpful, but ideally we could improve our classifier quality by down-sampling the training data to flatten the distribution of the target over each class.

There's a couple levels this problem exists at:
1. In automl, if we want to make a fair comparison between models, we should either use an automl objective which is hardened to imbalanced data (like Neyman-Pearson tracked by #457), or we should balance the data before computing the objective score.
2. During the training of each pipeline, each CV fold should be down-sampled to have a balanced distribution of labels. A simple way to do this is to compute the splits in automl as we do now, apply down-sampling to balance, and then train/evaluate each pipeline on each split. We could also have the pipelines generate the splits and down-sample internally, but if this is done right there wouldn't be any difference in accuracy.

I see [sklearn's doc mentions class imbalance](https://scikit-learn.org/stable/modules/cross_validation.html#cross-validation-iterators-with-stratification-based-on-class-labels), but I don't agree with their recommendation "to ensure that relative class frequencies is approximately preserved in each train and validation fold." I suppose we could put this to the test and see which strategy seems to generalize more effectively.

### PR DESCRIPTION
**Pull request by bchen1116**
part 1 of 2 for #973. 

This PR tracks adding in the custom sampler to EvalML, and we will turn this into a data splitter once merged with the [class imbalance PR](https://github.com/alteryx/evalml/pull/1775/). This does not include the actual data splitter itself.

Functionality of this sampler (and future data splitter) [here](https://alteryx.quip.com/Z1JbAwHFpoU3/Data-Splitting-Notes#aTMACAHdHbG)

### CODE REVIEW

File: evalml/preprocessing/data_splitters/balanced_classification_splitter.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
+import numpy as np
+
+from evalml.utils.woodwork_utils import (
+    _convert_woodwork_types_wrapper,
+    infer_feature_types
+)
+
+
+class BalancedClassificationSampler():
```
Comment (dsherry) [inline evalml/preprocessing/data_splitters/balanced_classification_splitter.py:9]: Missing base class

File: evalml/preprocessing/data_splitters/balanced_classification_splitter.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        normalized_counts = y.value_counts(normalize=True)
+        class_ratios = normalized_counts.copy().values
+        class_ratios /= class_ratios[-1]
+        # if no class ratios are greater than what we consider balanced, then the target is balanced
+        if all(class_ratios <= self.balanced_ratio):
+            return {}
+        # if any classes have than min_samples counts and are less than min_percentage of the total data,
+        # then it's severely imbalanced
+        if any(counts < self.min_samples) and any(normalized_counts < self.min_percentage):
+            return {}
```
Comment (freddyaboulton) [inline evalml/preprocessing/data_splitters/balanced_classification_splitter.py:61]: Playing devil's advocate here, should severely imbalanced and balanced data produce the same behavior?

The user's data can fall into four cases:
1. Balanced
2. Severely Imbalanced
3. Imbalanced and can downsample
4. Imbalanced but can't downsample because of `min_samples`

Right now users can't tell whether they are in cases 1, 2, or 4 and I think there is value in communicating that. Maybe it belongs in our class imbalance data check though.

File: evalml/tests/preprocessing_tests/test_balanced_classification_splitter.py [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+    pd.testing.assert_series_equal(y1, y2)
+
+
+@pytest.mark.parametrize("index", [[f'hello_{i}' for i in range(1000)],
+                                   random.shuffle([i + 0.5 for i in range(1000)]),
+                                   pd.MultiIndex.from_arrays([
+                                       [f"index_{i}" for i in range(1000)],
+                                       [i for i in range(1000)]
+                                   ])])
+def test_classification_imbalanced_custom_indices(index):
```
Comment (freddyaboulton) [inline evalml/tests/preprocessing_tests/test_balanced_classification_splitter.py:240]: 👍

File: evalml/preprocessing/data_splitters/balanced_classification_splitter.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+                If we don't need to resample, returns empty dictionary.
+        """
+        counts = y.value_counts()
+        normalized_counts = y.value_counts(normalize=True)
+        class_ratios = normalized_counts.copy().values
+        class_ratios /= class_ratios[-1]
+        # if no class ratios are greater than what we consider balanced, then the target is balanced
+        if all(class_ratios <= self.balanced_ratio):
+            return {}
+        # if any classes have than min_samples counts and are less than min_percentage of the total data,
```
Comment (freddyaboulton) [inline evalml/preprocessing/data_splitters/balanced_classification_splitter.py:58]: typo: "have less than min_samples"?

**Review by freddyaboulton — APPROVED** [review]
@bchen1116 I think this is great! The tests are very solid.

Nothing blocking from me.

File: evalml/preprocessing/data_splitters/balanced_classification_splitter.py (hunk 2 above)
Comment (bchen1116) [reply evalml/preprocessing/data_splitters/balanced_classification_splitter.py:61]: @freddyaboulton Yeah, I think in discussion with @dsherry, we needed to figure out how to get the class imbalance data check to work with this. Definitely agree that the behavior could be confusing, but I think as long as the data check can help us catch the severely-imbalanced case and warn the user, it clarifies the case.

Comment (bchen1116) [reply evalml/preprocessing/data_splitters/balanced_classification_splitter.py:61]: @freddyaboulton just filed a new [issue](https://github.com/alteryx/evalml/issues/1864) to track updating the class imbalance data check

**Review by chukarsten — APPROVED** [review]
Very nice job - super thorough testing.  Love it.  I had a comment that I don't *think* is merge blocking that we should probably resolve (or it is resolved already) concerning what to do when a class cannot be downsampled to the desired ratio.

File: evalml/tests/preprocessing_tests/test_balanced_classification_splitter.py [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+    assert y2.value_counts()[1] == 100
+
+
+def test_classification_data_drop():
+    # tests for whether or not the `max(0, counts[k] - goal_value)` code works as expected
+    X = pd.DataFrame({"a": [i for i in range(420)]})
+    y = pd.Series([0] * 90 + [1] * 100 + [2] * 120 + [3] * 40 + [4] * 70)
+    # will downsample the [2] target
+    # will try to downsample [0] and [4], but max(0, x) will prevent that
+    bcs = BalancedClassificationSampler(balanced_ratio=1, min_percentage=0.01)
```
[the remark this answers is not in the document]
Comment (bchen1116) [reply evalml/tests/preprocessing_tests/test_balanced_classification_splitter.py:340]: Mmm, yeah that's fair. My thought was that we cannot sample if we end up with less than `min_samples` values of a class. For instance, if we have 40 instances of class 1 and 100 instances of class 2, we might not want to sample the 100 instances down even if we had wanted a 1:1 ratio, whereas in a case of 40,000 vs 100,000 samples, it would be fine to sample the 100,000 to 40,000 cases since we have enough samples. As for warning or telling the user, I wasn't sure if this was something I should warn the user about or do silently. Thoughts @chukarsten @dsherry?

[the remark this answers is not in the document]
Comment (dsherry) [reply evalml/tests/preprocessing_tests/test_balanced_classification_splitter.py:340]: I think the understanding should be that if the number of instances falls below the threshold for certain classes, as it does for "0" and "4" here, that the method will simply not downsample those classes. No error in that case.
